samples/api/swapchain_recreation/swapchain_recreation.cpp

```diff
@@ -0,0 +1,1150 @@
+	// gets scheduled for destruction when the semaphore of the first image of the next
+	// swapchain can be recycled.
+	std::vector<PresentOperation> history_to_keep;
+	while (!present_history.empty())
+	{
+		PresentOperation &present_operation = present_history.back();
+
+		// If this is about an older swapchain, let it be.
+		if (present_operation.image_index == INVALID_IMAGE_INDEX)
+		{
+			assert(present_operation.cleanup_fence != VK_NULL_HANDLE);
+			break;
+		}
+
+		// Reset the index, so it's not processed in the future.
+		present_operation.image_index = INVALID_IMAGE_INDEX;
+
+		if (present_operation.cleanup_fence != VK_NULL_HANDLE)
+		{
+			// If there is already a fence associated with it, let it be cleaned up once
+			// the fence is signaled.
+			history_to_keep.push_back(std::move(present_operation));
+		}
+		else
+		{
+			assert(present_operation.present_semaphore != VK_NULL_HANDLE);
+
+			// Otherwise accumulate it in cleanup data.
+			cleanup.semaphores.push_back(present_operation.present_semaphore);
+
+			// Accumulate any previous swapchains that are pending destruction too.
+			for (SwapchainCleanupData &swapchain : present_operation.old_swapchains)
+			{
+				old_swapchains.emplace_back(swapchain);
+			}
+			present_operation.old_swapchains.clear();
+		}
+
+		present_history.pop_back();
+	}
+	std::move(history_to_keep.begin(), history_to_keep.end(), std::back_inserter(present_history));
+
+	if (cleanup.swapchain != VK_NULL_HANDLE || !cleanup.semaphores.empty())
+	{
+		old_swapchains.emplace_back(std::move(cleanup));
+	}
+}
+
+VkSemaphore SwapchainRecreation::get_semaphore()
+{
+	// If there is a free semaphore, return it
+	if (!semaphore_pool.empty())
+	{
+		VkSemaphore semaphore = semaphore_pool.back();
+		semaphore_pool.pop_back();
+		return semaphore;
+	}
+
+	VkSemaphore           semaphore = VK_NULL_HANDLE;
+	VkSemaphoreCreateInfo create_info{VK_STRUCTURE_TYPE_SEMAPHORE_CREATE_INFO};
+
+	VK_CHECK(vkCreateSemaphore(get_device_handle(), &create_info, nullptr, &semaphore));
+
+	return semaphore;
+}
+
+void SwapchainRecreation::recycle_semaphore(VkSemaphore semaphore)
+{
+	semaphore_pool.push_back(semaphore);
+}
+
+VkFence SwapchainRecreation::get_fence()
+{
+	// If there is a free fence, return it
+	if (!fence_pool.empty())
+	{
+		VkFence fence = fence_pool.back();
+		fence_pool.pop_back();
+		return fence;
+	}
+
+	VkFence           fence = VK_NULL_HANDLE;
+	VkFenceCreateInfo create_info{VK_STRUCTURE_TYPE_FENCE_CREATE_INFO};
+
+	VK_CHECK(vkCreateFence(get_device_handle(), &create_info, nullptr, &fence));
+
+	return fence;
+}
+
+void SwapchainRecreation::recycle_fence(VkFence fence)
+{
+	fence_pool.push_back(fence);
+
+	VK_CHECK(vkResetFences(get_device_handle(), 1, &fence));
+}
+
+VkPhysicalDevice SwapchainRecreation::get_gpu_handle()
+{
+	return get_device().get_gpu().get_handle();
+}
+
+VkDevice SwapchainRecreation::get_device_handle()
+{
+	return get_device().get_handle();
+}
+
+SwapchainRecreation::SwapchainRecreation()
```

Comment:
I'm too swamped right now, but maybe one day. The extension _is_ tested on Android both through CTS and ANGLE if you are more worried about the extension itself (and less about the sample)

Reply:
Hi @ShabbyX, is this sample still in the plans?  Thanks

Reply:
Yes, it was ready a year ago IMO. The TSG decided to wait for a driver release to test this with, and surprisingly that hasn't happened yet. I know @HansKristian-Work had an implementation for mesa last year, but that never landed? Nvidia surprisingly hasn't released a driver yet either.

So, I guess it'll just have to continue to wait. If an Android-savvy person likes to try this sample on Android (like Jose Emilion? :D) and give it the green light, by all means go ahead!

Reply:
Looking at my database, AMD and Nvidia support the extension:
https://vulkan.gpuinfo.org/listdevicescoverage.php?extension=VK_EXT_swapchain_maintenance1&platform=all

AMD is currently Linux only, nvidia is Windows and Linux :)

Reply:
Huh, I checked there, how did I miss that?

Ok well, like I said last year the TSG decided to wait until someone (other than myself) can try this sample on a desktop driver. So by all means, please do so it can land.

Reply:
Mesa supports Wayland and X11. X11 backend recently received support for present mode toggling as well.

NV's prop driver has some bugs with swapchain fence though, at least on Linux. A fix is supposedly on its way.

Reply:
OK, thanks.  We'll check it out and hopefully be close to merging by the next samples call..

Reply:
Just rebased, there was a minor conflict about .md->.adoc transition.

Also, running on SwiftShader right now, I see a new VVL error about the ANI semaphore not being unused. It triggers only once though, I'll have to look into it.

Reply:
Could be a VVL bug? As far as my logs show, the semaphore is not in use:

```
[info] ------------------------------------
[info] USAGE:
[info]  - Press v to enable v-sync (default)
[info]  - Press n to disable v-sync
[info]  - Press p to enable switching between compatible present modes (default)
[info]  - Press r to disable switching between compatible present modes
[info] Set environment variable USE_MAINTENANCE1=no to avoid VK_EXT_surface_maintenance1
[info] ------------------------------------
[warning] =======
[warning] NEW FRAME semaphore 0xa000000000a, fence 0x90000000009, present sem: 0xb000000000b
[warning] Calling ANI with semaphore 0xa000000000a
[warning] Submit, wait on 0xa000000000a, signal 0xb000000000b, signal fence 0x90000000009
[warning] Present, wait on 0xb000000000b
[warning] Present, signal fence 0xf000000000f
[warning] Clean up present fence -- fence 0xf000000000f is signaled
[warning] Clean up present fence -- recycle present semaphore 0xb000000000b
[warning] =======
[warning] NEW FRAME semaphore 0xb000000000b, fence 0xf000000000f, present sem: 0x100000000010
[warning] Calling ANI with semaphore 0xb000000000b
[error] [framework/core/instance.cpp:50] 1461184347 - VUID-vkAcquireNextImageKHR-semaphore-01779: Validation Error: [ VUID-vkAcquireNextImageKHR-semaphore-01779 ] Object 0: handle = 0xb000000000b, type = VK_OBJECT_TYPE_SEMAPHORE; | MessageID = 0x5717e75b | vkAcquireNextImageKHR():  Semaphore must not have any pending operations. The Vulkan spec states: If semaphore is not VK_NULL_HANDLE it must not have any uncompleted signal or wait operations pending (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkAcquireNextImageKHR-semaphore-01779)
[warning] Submit, wait on 0xb000000000b, signal 0x100000000010, signal fence 0xf000000000f
[warning] Present, wait on 0x100000000010
[warning] Present, signal fence 0x120000000012
[warning] Clean up present fence -- fence 0x120000000012 is signaled
[warning] Clean up present fence -- recycle present semaphore 0x100000000010
[warning] Recycle semaphore 0xa000000000a, fence 0x90000000009 (waited)
```

It's warning about the second call to ANI, which is using the semaphore used in the previous present operation. However, the fence for that semaphore is already signaled per "[warning] Clean up present fence -- fence 0xf000000000f is signaled". Maybe VVL is not accounting for VK_EXT_swapchain_maintenance1. I don't believe the app is doing anything wrong here. Diff used to reproduce these logs:

```
diff --git a/samples/api/swapchain_recreation/swapchain_recreation.cpp b/samples/api/swapchain_recreation/swapchain_recreation.cpp
index 1a5b29c..70c9e19 100644
--- a/samples/api/swapchain_recreation/swapchain_recreation.cpp
+++ b/samples/api/swapchain_recreation/swapchain_recreation.cpp
@@ -418,6 +418,7 @@ void SwapchainRecreation::setup_frame()
                vkWaitForFences(get_device_handle(), 1, &frame.submit_fence, true, UINT64_MAX);
 
                // Reset/recycle resources, they are no longer in use.
+               LOGW("Recycle semaphore {}, fence {} (waited)", (void *)frame.acquire_semaphore, (void *)frame.submit_fence);
                recycle_fence(frame.submit_fence);
                recycle_semaphore(frame.acquire_semaphore);
                vkResetCommandPool(get_device_handle(), frame.command_pool, 0);
@@ -443,6 +444,8 @@ void SwapchainRecreation::setup_frame()
        frame.submit_fence      = get_fence();
        frame.acquire_semaphore = get_semaphore();
        frame.present_semaphore = get_semaphore();
+       LOGW("=======");
+       LOGW("NEW FRAME semaphore {}, fence {}, present sem: {}", (void *)frame.acquire_semaphore, (void *)frame.submit_fence, (void *)frame.present_semaphore);
 
        if (frame.command_pool == VK_NULL_HANDLE)
        {
@@ -542,6 +545,7 @@ void SwapchainRecreation::render(uint32_t index)
        info.pWaitDstStageMask    = &wait_stage;
        info.signalSemaphoreCount = 1;
        info.pSignalSemaphores    = &frame.present_semaphore;
+       LOGW("Submit, wait on {}, signal {}, signal fence {}", (void *)frame.acquire_semaphore, (void *)frame.present_semaphore, (void *)frame.submit_fence);
        VK_CHECK(vkQueueSubmit(queue->get_handle(), 1, &info, frame.submit_fence));
 }
 
@@ -576,6 +580,7 @@ VkResult SwapchainRecreation::acquire_next_image(uint32_t *index)
        // There is no need for this with VK_EXT_swapchain_maintenance1.
        VkFence acquire_fence = has_maintenance1 ? VK_NULL_HANDLE : get_fence();
 
+       LOGW("Calling ANI with semaphore {}", (void *)frame.acquire_semaphore);
        VkResult result = vkAcquireNextImageKHR(get_device_handle(), swapchain, UINT64_MAX, frame.acquire_semaphore, acquire_fence, index);
 
        if (has_maintenance1 && (result == VK_SUCCESS || result == VK_SUBOPTIMAL_KHR))
@@ -623,6 +628,7 @@ VkResult SwapchainRecreation::present_image(uint32_t index)
        present.pImageIndices      = &index;
        present.waitSemaphoreCount = 1;
        present.pWaitSemaphores    = &frame.present_semaphore;
+       LOGW("Present, wait on {}", (void *)frame.present_semaphore);
 
        // When VK_EXT_swapchain_maintenance1 is enabled, add a fence to the present operation,
        // which is signaled when the resources associated with present operation can be freed.
@@ -635,6 +641,7 @@ VkResult SwapchainRecreation::present_image(uint32_t index)
 
                fence_info.swapchainCount = 1;
                fence_info.pFences        = &present_fence;
+               LOGW("Present, signal fence {}", (void *)present_fence);
 
                present.pNext = &fence_info;
 
@@ -710,6 +717,7 @@ void SwapchainRecreation::cleanup_present_history()
                }
                VK_CHECK(result);
 
+               LOGW("Clean up present fence -- fence {} is signaled", (void *)present_info.cleanup_fence);
                cleanup_present_info(present_info);
                present_history.pop_front();
        }
@@ -752,6 +760,7 @@ void SwapchainRecreation::cleanup_present_info(PresentOperationInfo &present_inf
        // clean up.  That fence is placed in the present history just for clean up purposes.
        if (present_info.present_semaphore != VK_NULL_HANDLE)
        {
+               LOGW("Clean up present fence -- recycle present semaphore {}", (void *)present_info.present_semaphore);
                recycle_semaphore(present_info.present_semaphore);
        }
```

Reply:
I noticed in the meeting notes that this sample is "Waiting on updates". There are no updates coming, this was confirmed to be a VVL gap (https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/7025). Unless you are waiting for the VVL error to go away, that's blocked on that issue. If not, the change is good to land.

Reply:
The VVL bug has been fixed. I just rebased this to retrigger the CI.